File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        n=len(preorder)
        if len(inorder)!=len(preorder):
            return None
        dic={}
        for i in range(n):
            dic[inorder[i]]=i
        istart=0
        iend=len(inorder)-1
        pstart=0
        pend=len(preorder)-1
        def build(inorder,preorder,istart,iend,pstart,pend):
            if istart>iend or pstart>pend:
                return None
            root=TreeNode(preorder[pstart])
            isroot=dic[preorder[pstart]]
            numleft=isroot-istart
            root.left=build(inorder,preorder,istart,isroot-1,pstart+1,pstart+numleft)
            root.right=build(inorder,preorder,isroot+1,iend,pstart+numleft+1,pend)
            return root
        root=build(inorder,preorder,istart,iend,pstart,pend)
        return root

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* construct(vector<int>&Inorder,vector<int>&Preorder,int prein,int preout,int In,int Inout,unordered_map<int,int>&mp){
        if(prein>preout and In>Inout)return NULL;
        TreeNode* root=new TreeNode(Preorder[prein]);
        int numsLeft=mp[root->val]-In;   
        root->left=construct(Inorder,Preorder,prein+1,prein+numsLeft,In,mp[root->val]-1,mp);
        root->right=construct(Inorder,Preorder,prein+numsLeft+1,preout,mp[root->val]+1,Inout,mp);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int,int>mp;
        int n=inorder.size();
        for(int i=0;i<n;i++){
            mp[inorder[i]]=i;
        }
        return construct(inorder,preorder,0,preorder.size()-1,0,inorder.size()-1,mp);
        //return root;
    }
};